### PR TITLE
doc: autogenerate python binding docs with Sphinx

### PR DESCRIPTION
**Pull request by SteVwonder**
Use Sphinx's napoleon to pre-process Google style docstrings into ReSt
and then use Sphinx's autodoc to import the flux py modules and
autogenerate documentation

### CODE REVIEW

**Comment by grondo** [comment]
@SteVwonder, I probably need some simple instructions here, but I'm getting a lot of warnings when sphinx tries to generate the python docs. I get a `doc/python` directory, but there's not much content in the files there, and none of the python docs appear in `html`:

Here's the entirety of `python/flux.job.Jobspec.rst`:

```rst
flux.job.Jobspec module
=======================

.. automodule:: flux.job.Jobspec
   :members:
   :undoc-members:
   :show-inheritance:

```

This is just a sample of warnings I'm seeing:
```
WARNING: autodoc: failed to import module 'hostlist' from module 'flux'; the following exception was raised:
No module named 'flux.hostlist'
WARNING: autodoc: failed to import module 'idset' from module 'flux'; the following exception was raised:
No module named 'flux.idset'
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:8: WARNING: Unexpected indentation.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:6: WARNING: Inline literal start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:6: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:10: WARNING: Unexpected indentation.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:15: WARNING: Block quote ends without a blank line; unexpected unindent.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:18: WARNING: Unexpected indentation.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:23: WARNING: Bullet list ends without a blank line; unexpected unindent.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:23: WARNING: Inline literal start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_counts:23: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:12: WARNING: Unexpected indentation.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:10: WARNING: Inline literal start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:10: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:14: WARNING: Unexpected indentation.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:16: WARNING: Block quote ends without a blank line; unexpected unindent.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:16: WARNING: Inline literal start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resource_walk:16: WARNING: Inline interpreted text or phrase reference start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.event_watch:5: WARNING: Unexpected indentation.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.event_watch:11: WARNING: Inline emphasis start-string without end-string.
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.event_watch_async:12: WARNING: Inline emphasis start-string without end-string.
WARNING: autodoc: failed to import module 'JobID' from module 'flux.job'; the following exception was raised:
No module named 'flux.job.JobID'; 'flux.job' is not a package
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec:1: WARNING: duplicate object description of flux.job.Jobspec, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.attributes:1: WARNING: duplicate object description of flux.job.Jobspec.attributes, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.cwd:1: WARNING: duplicate object description of flux.job.Jobspec.cwd, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.duration:1: WARNING: duplicate object description of flux.job.Jobspec.duration, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.environment:1: WARNING: duplicate object description of flux.job.Jobspec.environment, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.resources:1: WARNING: duplicate object description of flux.job.Jobspec.resources, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.tasks:1: WARNING: duplicate object description of flux.job.Jobspec.tasks, other instance in python/flux.job, use :noindex: for one of them
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.Jobspec.version:1: WARNING: duplicate object description of flux.job.Jobspec.version, other instance in python/flux.job, use :noindex: for one of them
WARNING: autodoc: failed to import module 'event' from module 'flux.job'; the following exception was raised:
No module named 'flux.job.event'; 'flux.job' is not a package
WARNING: autodoc: failed to import module 'info' from module 'flux.job'; the following exception was raised:
No module named 'flux.job.info'; 'flux.job' is not a package
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.kill:1: WARNING: duplicate object description of flux.job.kill, other instance in python/flux.job, use :noindex: for one of them
WARNING: autodoc: failed to import module 'kvs' from module 'flux.job'; the following exception was raised:
No module named 'flux.job.kvs'; 'flux.job' is not a package
WARNING: autodoc: failed to import module 'list' from module 'flux.job'; the following exception was raised:
No module named 'flux.job.list'; 'flux.job' is not a package
/usr/lib/python3/dist-packages/flux/job.py:docstring of flux.job.submit:1: WARNING: duplicate object description of flux.job.submit, other instance in python/flux.job, use :noindex: for one of them
```

**Comment by grondo** [comment]
Ok, my problem is obvious from the paths in the warnings above. Sphinx is reading the system installed Flux python bindings, instead of the srcdir. I think the problem is that the project path is appended instead of prepended to `sys.path`:

```diff
 grondo@asp:~/git/flux-core.git/doc$ git diff
diff --git a/doc/conf.py b/doc/conf.py
index 36cebf428..1dee3308f 100644
--- a/doc/conf.py
+++ b/doc/conf.py
@@ -66,7 +66,7 @@ except:
     build_dir = script_dir
 
 # Make sure that the python bindings are in PYTHONPATH for autodoc
-sys.path.append(py_bindings_dir)
+sys.path.insert(0, py_bindings_dir)
 
 # run api doc
 def run_apidoc(_):

```

However, I still don't seem to be getting any actual content in `python/*.rst` nor in `html`.

**Comment by SteVwonder** [comment]
> Here's the entirety of python/flux.job.Jobspec.rst:

Surprisingly, that is the expected content and all that is required.  Sphinx will use that to autogenerate html with the docstrings that it picks up from importing the bindings.  That just tells it what information to pull in and render. Down the road, if we want to add text/context to the bindings or show more/less auto-generated information, we can check the auto-generated python/*.rst files into version control.  For now, I figured it was more convenient to have Sphinx build them automatically so that they are in sync with the bindings.

> Sphinx is reading the system installed Flux python bindings, instead of the srcdir. I think the problem is that the project path is appended instead of prepended to sys.path:

Thanks @grondo, for running that down! I pushed a fixup commit with that change.

> However, I still don't seem to be getting any actual content in python/*.rst nor in html.

I waffled on whether or not to generate html as part of the normal build.  I figured it wouldn't be all that useful to have html files installed with the man pages, but I can definitely see how it would be useful to quickly generate html to see RTD output before pushing to GitHub.  I just pushed a commit that adds an `html` make target for convenience.  Running `make html` in the `doc` dir should now produce an `html` directory with the content you are expecting.

Let me know if you think it would be useful to build the html pages as part of the normal build or if you have any suggestions on how to make building the html out-of-band easier/more obvious.

**Comment by grondo** [comment]
Awesome! Thanks, I'll take a look tomorrow.

File: doc/Makefile.am [hunk 1]
```diff
@@ -285,18 +285,33 @@ sphinx_verbose = $(sphinx_verbose_$(V))
...
+		$(STDERR_DEVNULL)
 	@echo "  MV        manpages"; \
 	for sec in 1 3 5 7; do \
 	  $(MKDIR_P) man$$sec && \
 	  mv -f $(abs_builddir)/man/*.$$sec man$$sec/; \
 	done
 
+.PHONY: html
+html: conf.py $(RST_FILES)
+	SPHINX_BUILDDIR=$(abs_builddir) $(PYTHON) \
```
Comment (grondo) [inline doc/Makefile.am:311]: Prefix with `$(sphinx_verbose)`?

**Review by grondo — APPROVED** [review]
Great! This is working for me now. I think `make html` down in docs dir is fine for now. We can always modify if the future if it doesn't meet developer's needs.

Just one minor suggestion inline.

Thanks!

File: doc/Makefile.am (hunk 1 above)
Comment (SteVwonder) [reply doc/Makefile.am:311]: That variable is set to echo `BUILD manpages`, which is a bit of a misnomer for the html.  Should I create a new variable that echos `BUILD html`?

Comment (grondo) [reply doc/Makefile.am:311]: Ah, good point. You can do that or leave as is - up to you. It isn't too important...

Comment (SteVwonder) [reply doc/Makefile.am:316]: Sounds good.  Just did a squash of my old fixup commit and amended the commit at the top of the commit stack (which added the `html` target) to include a new variable that echos `BUILD html`.

**Comment by SteVwonder** [comment]
I completely forgot the fact that @gonsie already did most of this on #3123 🤦 .  There are quite a few things in the docstrings that are better in that PR.  Let me cherry-pick a few of those commits onto this PR and add some co-authorship to the existing commits.  In particular, I was unaware of that `..code-block` could be used in the docstrings.  That is much nicer than what I did.

**Comment by grondo** [comment]
> I completely forgot the fact that @gonsie already did most of this on #3123 ￼ . 

Sounds good, sorry I missed that too!

**Comment by SteVwonder** [comment]
Just pushed those changes/cherry-picks.  This should be good.  @grondo: would you like to take another look?  If not, am I good to set MWP?

**Comment by grondo** [comment]
Can you add a commit body to 44a9369 (the other commits have nicely descriptive commit messages, and it will probably help us in the future if this one did too)

Otherwise it looks good to me!

**Comment by SteVwonder** [comment]
> Can you add a commit body

Sure! Just pushed.  I'll set MWP so this can go in once the CI runs again.

**Pull request merged**